cloudinit/event.py

```diff
@@ -1,17 +1,72 @@
 # This file is part of cloud-init. See LICENSE file for license information.
-
 """Classes and functions related to event handling."""
 
+import copy
+
+from cloudinit import log as logging
+from cloudinit import util
+
+LOG = logging.getLogger(__name__)
+
 
-# Event types which can generate maintenance requests for cloud-init.
 class EventType(object):
+    """Event types which can generate maintenance requests for cloud-init."""
+
     BOOT = "System boot"
     BOOT_NEW_INSTANCE = "New instance first boot"
 
-    # TODO: Cloud-init will grow support for the follow event types:
-    # UDEV
+    # Cloud-init should grow support for the follow event types:
+    # HOTPLUG
     # METADATA_CHANGE
     # USER_REQUEST
 
 
+EventTypeMap = {
+    'boot': EventType.BOOT,
+    'boot-new-instance': EventType.BOOT_NEW_INSTANCE,
+}
+
+
+# inverted mapping
+EventNameMap = {v: k for k, v in EventTypeMap.items()}
+
+
+def get_enabled_events(config_events: dict, default_events: dict) -> dict:
+    """Determine which update events are allowed.
+
+    Merge datasource capabilities with system config to determine events
+    """
+    # updates:
+    #   network:
+    #     when: [boot]
+
+    LOG.debug('updates user: %s', config_events)
+    LOG.debug('updates default: %s', default_events)
+
+    # If a key exists in multiple places, the first in the list wins.
+    updates = util.mergemanydict([
+        copy.deepcopy(config_events),
+        copy.deepcopy(default_events),
+    ])
+    LOG.debug('updates merged: %s', updates)
+
+    events = {}
+    for etype in [scope for scope, val in default_events.items()
+                  if type(val) == dict and 'when' in val]:
+        events[etype] = (
+            set([EventTypeMap.get(evt)
+                 for evt in updates.get(etype, {}).get('when', [])
+                 if evt in EventTypeMap]))
```

Comment:
Our codebase has a bunch of places where we pass around dicts with a known set of attributes which have a bunch of semantics attached to them, and it's always difficult to follow what's going on because the semantics are defined wherever that dict happens to be used in the codebase. The complexity of these list comprehensions makes me wonder if we should be replacing/wrapping these dicts with a class which could encapsulate some of this complexity, to avoid going down the same path that we have before?

Something like (this very quick sketch):

```py
updates = UpdateEventsConfig.from_merge(config_events, default_events)
...
for scope in default_events.scopes:
    events[scope] = updates.events_for(scope)
```

is much easier to follow, for example.

(Potentially a good first step would be to split out each of these list comprehensions to functions: giving them a name and explanatory docstring/comment might be sufficiently clarifying to obviate the need for a class.)